pkg/flare/writer.go

```diff
@@ -95,7 +96,11 @@ func (f *RedactingWriter) Write(p []byte) (int, error) {
 		n, err = f.target.Write(cleaned)
 	}
 
-	return n, err
+	if n != len(cleaned) {
+		err = io.ErrShortWrite
+	}
+
+	return len(p), err
```

Comment:
This is tricky from an interface implementation standpoint.... we're passing `p []byte`, but we're really writing `cleaned`, and I'm not sure all the regex cleanup will guarantee `len(p)==len(cleaned)` (probably not). Is it therefore accurate to return `len(p)` when `n != len(cleaned)` (and we have a `ShortWrite`)? I don't think that's strictly true, but not easy to get around.

Reply:
`len(p) != len(cleaned)` quite often. That's why returning `len(cleaned)` is wrong. The caller of `Write` expects to know how many bytes were _processed_ by this Writter interface, even if what was actually written is a single byte (or none, like in io.Discard). 

This is in a similar way to the `write` system call in C: as the caller, you expect it to return the length of the buffer you passed in, regardless of the actual side effects the write function has.

It's not obvious, but I asked a Go language developer about this and this is the right way.

About the extra if I added: It is just to detect the same error that Copy was detecting on us, on the underlying Writter.

Reply:
I definitely prefer this new behavior, it's better and less misleading to what we previously had. 
Write processes the `p` buffer, it writes what it can, and returns the number of bytes redacted/processed (not written). And we can maybe use the error returned somehow in the called to make a decision. 

That said, this isn't strictly respecting the `Writer` interface (it was previously broken too!). Let me clarify what I mean. As you say `len(p) != len(cleaned)` is probably the most common case, as redacting the input will result in a different length byte slice. What I'm saying is that from an interface perspective that makes this `Write` function hard to use. Didn't mean to imply this was your fault, it's a design problem stemming from adding the redacting call inside the `Write()` function. Let me underline that this was broken from day 1, nothing to do with your PR.

The C `write()` syscall returns the number of bytes written:
```
RETURN VALUES
     Upon successful completion the number of bytes which were written is returned.  Otherwise, a -1 is returned and the global variable errno is set to indicate the error.
```

Similarly, the go `Write` interface is described as:
```
Write writes len(p) bytes from p to the underlying data stream. It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early. Write must return a non-nil error if it returns n < len(p). Write must not modify the slice data, even temporarily.

type Writer interface {
        Write(p []byte) (n int, err error)
}
```

Since what we actually write is different to what the user passes, and the user has no idea what `len(redacted)` is, the value we return for bytes written would him nothing about how much of the `p` buffer passed in was written, which I believe is the goal of the return value. So I understand your change is maybe less prone to confusion, because it's an all or nothing approach - whereas the previous implementation was downright returning a meaningless number of bytes as far as the input buffer went. That's the part I meant was _"not easy to get around"_: implementing a return value that would accurately reflect the number of bytes from buffer `p` that were actually written (when writing the `redacted` buffer). Currently we say the entire `p` slice was written in the event of a short write, even if that wasn't true. And this might be the most reasonable thing we can do.

The only way I feel we could do better is if `cleaned, err := log.CredentialsCleanerBytes(p)` guaranteed `len(p) == len(cleaned)`, then we could return n.

Reply:
Just a note about C's `write()` and Go's `Write` interface: when they talk about "number of bytes which were written", it is from the caller perspective.